src/MemoryTracing/MemoryTracingIntegrationTest.cpp

```diff
@@ -0,0 +1,145 @@
+#include <gtest/gtest.h>
+
+#include "GrpcProtos/Constants.h"
+#include "MemoryTracing/MemoryInfoListener.h"
+#include "MemoryTracing/MemoryInfoProducer.h"
+#include "OrbitBase/Logging.h"
+#include "capture.pb.h"
+
+namespace orbit_memory_tracing {
+namespace {
+
+using orbit_grpc_protos::kMissingInfo;
+using orbit_grpc_protos::ProducerCaptureEvent;
+using orbit_grpc_protos::SystemMemoryUsage;
+using orbit_memory_tracing::MemoryInfoListener;
+using orbit_memory_tracing::MemoryInfoProducer;
+
+class BufferMemoryInfoListener : public MemoryInfoListener {
+ public:
+  void OnSystemMemoryUsage(SystemMemoryUsage system_memory_usage) override {
+    ProducerCaptureEvent event;
+    *event.mutable_system_memory_usage() = std::move(system_memory_usage);
+    {
+      absl::MutexLock lock{&events_mutex_};
+      events_.emplace_back(std::move(event));
+    }
+  }
+
+  [[nodiscard]] std::vector<ProducerCaptureEvent> GetAndClearEvents() {
+    absl::MutexLock lock{&events_mutex_};
+    std::vector<ProducerCaptureEvent> events = std::move(events_);
+    events_.clear();
+    return events;
+  }
+
+ private:
+  std::vector<ProducerCaptureEvent> events_;
+  absl::Mutex events_mutex_;
+};
+
+class MemoryTracingIntegrationTestFixture {
+ public:
+  MemoryTracingIntegrationTestFixture(uint64_t memory_sampling_period_ns)
+      : memory_sampling_period_ns_(memory_sampling_period_ns) {}
+
+  void StartTracingAndWaitForTracingLoopStarted() {
+    CHECK(!producer_.has_value());
+    CHECK(!listener_.has_value());
+    producer_.emplace(memory_sampling_period_ns_);
+    listener_.emplace();
+    producer_->SetListener(&*listener_);
+    producer_->Start();
+  }
+
+  [[nodiscard]] std::vector<ProducerCaptureEvent> StopTracingAndGetEvents() {
+    CHECK(producer_.has_value());
+    CHECK(listener_.has_value());
+    producer_->Stop();
+    producer_.reset();
+    std::vector<ProducerCaptureEvent> events = listener_->GetAndClearEvents();
+    listener_.reset();
+    return events;
+  }
+
+ private:
+  uint64_t memory_sampling_period_ns_;
+  std::optional<MemoryInfoProducer> producer_ = std::nullopt;
+  std::optional<BufferMemoryInfoListener> listener_ = std::nullopt;
+};
+
+[[nodiscard]] std::vector<ProducerCaptureEvent> TraceAndGetEvents(
+    MemoryTracingIntegrationTestFixture* fixture) {
+  CHECK(fixture != nullptr);
+
+  fixture->StartTracingAndWaitForTracingLoopStarted();
+
+  constexpr absl::Duration kSleepBeforeStopTracing = absl::Milliseconds(1000);
+  absl::SleepFor(kSleepBeforeStopTracing);
+  return fixture->StopTracingAndGetEvents();
+}
+
+void VerifyOrderAndContentOfEvents(const std::vector<ProducerCaptureEvent>& events,
+                                   uint64_t sampling_period_ns) {
+  const uint64_t kSamplingTimeErrorToleranceNs = absl::Milliseconds(1) / absl::Nanoseconds(1);
+  uint64_t previous_event_timestamp_ns = 0;
+  for (const auto& event : events) {
+    if (event.event_case() != ProducerCaptureEvent::kSystemMemoryUsage) UNREACHABLE();
```

Comment:
It's strange to use `UNREACHABLE` like this, that's more for branches of `switch`es and `if`s that cannot be avoided.
In this case I would just write
`CHECK(event.event_case() == ProducerCaptureEvent::kSystemMemoryUsage)`

Reply:
Make sense! Changed accordingly. Thanks!